### PR TITLE
feat: supply string type

### PR DESCRIPTION
**Pull request by hongda3141**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-modbus-go/blob/master/.github/Contributing.md

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by hongda3141** [comment]
change log:
1. modify  variable "stringLength" to "wordLength"
2. change bit length of string from 1 to 16
3. change swith-case to if-else block

**Comment by hongda3141** [comment]
@weichou1229 
PTAL again ，thanks

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]
We changed the DeviceProfile data model.
Please modify the sample profile.  Here is the example:
https://github.com/edgexfoundry/device-modbus-go/blob/b09d9ba1a72cafb6b38617a4d5eb8b9f0ac4b1bc/cmd/res/profiles/modbus.test.device.profile.yml

File: internal/driver/deviceclient.go [hunk 1]
```diff
@@ -183,7 +195,23 @@ func TransformCommandValueToDataBytes(commandInfo *CommandInfo, value *models.Co
...
+			less := make([]byte, oribytel-bytesl)
+			dataBytes = append(tempBytes, less...)
+		} else if bytesl > oribytel {
+			dataBytes = tempBytes[:oribytel]
+		} else {
+			dataBytes = []byte(oriStr)
+		}
+	} else {
+		dataBytes = numericValue[maxSize-byteCount : maxSize]
+	}
```
Comment (weichou1229) [inline internal/driver/deviceclient.go:214]: Please move those code snippet to https://github.com/edgexfoundry/device-modbus-go/blob/700bb1d367cb9371150de4b976218c3919e9a955/internal/driver/deviceclient.go#L260

**Review by weichou1229 — CHANGES_REQUESTED** [review]

File: internal/driver/deviceclient.go [hunk 2]
```diff
@@ -183,7 +195,23 @@ func TransformCommandValueToDataBytes(commandInfo *CommandInfo, value *models.Co
...
+		oribytel := int(commandInfo.Length * 2)
+		if bytesl < oribytel {
+			less := make([]byte, oribytel-bytesl)
+			dataBytes = append(tempBytes, less...)
+		} else if bytesl > oribytel {
+			dataBytes = tempBytes[:oribytel]
+		} else {
+			dataBytes = []byte(oriStr)
+		}
+	} else {
```
Comment (weichou1229) [inline internal/driver/deviceclient.go:212]: Please move those code snippet to https://github.com/edgexfoundry/device-modbus-go/blob/700bb1d367cb9371150de4b976218c3919e9a955/internal/driver/deviceclient.go#L260

**Review by weichou1229 — CHANGES_REQUESTED** [review]

File: internal/driver/deviceclient.go (hunk 2 above)
Comment (weichou1229) [inline internal/driver/deviceclient.go:212]: Please move those code snippet to https://github.com/edgexfoundry/device-modbus-go/blob/700bb1d367cb9371150de4b976218c3919e9a955/internal/driver/deviceclient.go#L260

**Review by weichou1229 — CHANGES_REQUESTED** [review]

File: internal/driver/constant.go [hunk 3]
```diff
@@ -36,6 +36,8 @@ const (
 	// RAW_TYPE define binary data type which read from Modbus device
 	RAW_TYPE = "rawType"
 
+	// WORD_LENGTH  E.g. "abcd" need 4 bytes as is 2 word, so WORD_LENGTH=2
+	WORD_LENGTH            = "wordLength"
```
Comment (weichou1229) [inline internal/driver/constant.go:40]: Rename the `wordLength` to `stringRegisterSize` and check the register size should be within the range of 1~123.

The register size limit can refer to https://modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf
> 6.12 16 (0x10) Write Multiple registers
This function code is used to write a block of contiguous registers (1 to 123 registers) in a
remote device.

**Review by weichou1229 — CHANGES_REQUESTED** [review]

**Comment by hongda3141** [comment]
@weichou1229
PTAL again ，thanks a lot for the advice

File: cmd/res/devices/modbus.test.devices.toml [hunk 4]
```diff
@@ -31,3 +31,20 @@
...
+  labels = [ 'modbus TCP' ]
+  [DeviceList.Protocols]
+    [DeviceList.Protocols.modbus-tcp]
+        Address = '0.0.0.0'
+        Port = '1502'
+        UnitID = '1'
+  [[DeviceList.AutoEvents]]
+        Frequency = '20s'
+        OnChange = false
+        Resource = 'ReadString'
```
Comment (weichou1229) [inline cmd/res/devices/modbus.test.devices.toml:49]: This pre-defined device is invalid. 
The valid format should be: 
```
[[DeviceList]]
  Name = 'Modbus-TCP-Read-String'
  ProfileName = 'Test.Device.Modbus.String.Profile'
  Description = 'use for auto read a string value'
  labels = [ 'modbus TCP' ]
  [DeviceList.Protocols]
    [DeviceList.Protocols.modbus-tcp]
        Address = '0.0.0.0'
        Port = '1502'
        UnitID = '1'
        Timeout = '5'
        IdleTimeout = '5'
  [[DeviceList.AutoEvents]]
        Interval = '20s'
        OnChange = false
        SourceName = 'ReadString'
```

File: cmd/res/profiles/modbus.test.device.string.profile.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+labels:
+  - "no lable"
+description: "this profile is testing  string value type with read and write"
+
+deviceResources:
+  -
+    name: "StringA"
+    description: "no description"
+    attributes:
+      { primaryTable: "INPUT_REGISTERS", startingAddress: "1", stringRegisterSize: "1"}
```
Comment (weichou1229) [inline cmd/res/profiles/modbus.test.device.string.profile.yml:13]: The startingAddress and stringRegisterSize should be integer. Please change to `{ primaryTable: "INPUT_REGISTERS", startingAddress: 1, stringRegisterSize: 1}`

File: cmd/res/profiles/modbus.test.device.string.profile.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    properties:
+      valueType: "String"
+      readWrite: "R"
+
+  -
+    name: "StringB"
+    description: "no description"
+    attributes:
+      { primaryTable: "HOLDING_REGISTERS", startingAddress: "15", stringRegisterSize: "5"}
```
Comment (weichou1229) [inline cmd/res/profiles/modbus.test.device.string.profile.yml:22]: The startingAddress and stringRegisterSize should be integer. Please change to `{ primaryTable: "INPUT_REGISTERS", startingAddress: 1, stringRegisterSize: 1}`

**Review by weichou1229 — CHANGES_REQUESTED** [review]
No problem.
Since we change the data model in V2, please verify the GET and PUT command works with edgex-go master branch.

**Comment by hongda3141** [comment]
@weichou1229
that`s right, profiles should update to v2 too,
PTAL again ，thanks

**Comment by weichou1229** [comment]
@hongda3141 
The Get and Put command works incorrectly.

1. For the Get command, I thought the value should be decoded to a readable string.
`"value" : "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"`
```
% curl http://localhost:59882/api/v2/device/name/Modbus-TCP-Read-String/StringB | json_pp
{
   "apiVersion" : "v2",
   "event" : {
      "apiVersion" : "v2",
      "deviceName" : "Modbus-TCP-Read-String",
      "id" : "669f5d9f-2004-47ad-9533-e15ff452304d",
      "origin" : 1623834796385661000,
      "profileName" : "Test.Device.Modbus.String.Profile",
      "readings" : [
         {
            "binaryValue" : null,
            "deviceName" : "Modbus-TCP-Read-String",
            "id" : "3fb25504-d6da-4525-8143-3a55f84299ad",
            "mediaType" : "",
            "origin" : 1623834796385624000,
            "profileName" : "Test.Device.Modbus.String.Profile",
            "resourceName" : "StringB",
            "value" : "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000",
            "valueType" : "String"
         }
      ],
      "sourceName" : "StringB"
   },
   "statusCode" : 200
}
```

2. For the Put command, I got the transform error.
```
% curl http://localhost:59882/api/v2/device/name/Modbus-TCP-Read-String/StringB -X PUT -H "Content-Type: application/json" -d '{
    "StringB":"test-string"
  }' | json_pp
{
   "apiVersion" : "v2",
   "message" : "request failed, status code: 500, err: {\"apiVersion\":\"v2\",\"message\":\"error writing DeviceResourece StringB for Modbus-TCP-Read-String -\\u003e transform command value failed, err: failed to transform test-string to []byte\",\"statusCode\":500}",
   "statusCode" : 500
}

```

**Comment by hongda3141** [comment]
> @hongda3141
> The Get and Put command works incorrectly.
> 
> 1. For the Get command, I thought the value should decode to a readable string.
>    `"value" : "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"`
> 
> ```
> % curl http://localhost:59882/api/v2/device/name/Modbus-TCP-Read-String/StringB | json_pp
> {
>    "apiVersion" : "v2",
>    "event" : {
>       "apiVersion" : "v2",
>       "deviceName" : "Modbus-TCP-Read-String",
>       "id" : "669f5d9f-2004-47ad-9533-e15ff452304d",
>       "origin" : 1623834796385661000,
>       "profileName" : "Test.Device.Modbus.String.Profile",
>       "readings" : [
>          {
>             "binaryValue" : null,
>             "deviceName" : "Modbus-TCP-Read-String",
>             "id" : "3fb25504-d6da-4525-8143-3a55f84299ad",
>             "mediaType" : "",
>             "origin" : 1623834796385624000,
>             "profileName" : "Test.Device.Modbus.String.Profile",
>             "resourceName" : "StringB",
>             "value" : "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000",
>             "valueType" : "String"
>          }
>       ],
>       "sourceName" : "StringB"
>    },
>    "statusCode" : 200
> }
> ```
> 
> 1. For the Put command, I got the transform error.
> 
> ```
> % curl http://localhost:59882/api/v2/device/name/Modbus-TCP-Read-String/StringB -X PUT -H "Content-Type: application/json" -d '{
>     "StringB":"test-string"
>   }' | json_pp
> {
>    "apiVersion" : "v2",
>    "message" : "request failed, status code: 500, err: {\"apiVersion\":\"v2\",\"message\":\"error writing DeviceResourece StringB for Modbus-TCP-Read-String -\\u003e transform command value failed, err: failed to transform test-string to []byte\",\"statusCode\":500}",
>    "statusCode" : 500
> }
> ```

thanks very much，i`ll fix that

**Comment by hongda3141** [comment]
> No problem.
> Since we change the data model in V2, please verify the GET and PUT command works with edgex-go master branch.

@weichou1229  hi，i found there is no edgex-lanch.sh in branch master, could you tell me which branch did you used for edgex-go? i always used and tested in fuji, but i think maybe it is too old for v2 version.

**Comment by weichou1229** [comment]
@hongda3141 

> i found there is no edgex-lanch.sh in branch master

It was removed in master branch

>  could you tell me which branch did you used for edgex-go?

master branch


You can use the following script before community add it back.
```
#!/bin/bash

DIR=$PWD
CMD=../cmd

# Kill all edgex-* stuff
function cleanup {
	pkill edgex
}

# disable secret-store integration
export EDGEX_SECURITY_SECRET_STORE=false

###
# Core Command
###
cd $CMD/core-command
# Add `edgex-` prefix on start, so we can find the process family
exec -a edgex-core-command ./core-command &
cd $DIR

###
# Core Data
###
cd $CMD/core-data
exec -a edgex-core-data ./core-data &
cd $DIR

###
# Core Metadata
###
cd $CMD/core-metadata
exec -a edgex-core-metadata ./core-metadata &
cd $DIR

trap cleanup EXIT

while : ; do sleep 1 ; done
```

**Comment by hongda3141** [comment]
thanks a lot

**Comment by hongda3141** [comment]
@weichou1229
i fixed the PUT and GET command error, 
PTAL thanks

**Review by weichou1229 — CHANGES_REQUESTED** [review]

File: internal/driver/deviceclient.go [hunk 7]
```diff
@@ -157,6 +167,8 @@ func TransformDataBytesToResult(req *models.CommandRequest, dataBytes []byte, co
 		if (dataBytes[0] & 1) > 0 {
 			res = true
 		}
+	case common.ValueTypeString:
+		res = trimString(dataBytes)
```
[the remark this answers is not in the document]
Comment (hongda3141) [reply internal/driver/deviceclient.go]: yeah , that is more concise than writing another function

**Comment by hongda3141** [comment]
@weichou1229
i update the codes with using build-in method for trimming the zero value, 
PTAL thanks

**Review by weichou1229 — APPROVED** [review]
LGTM

**Comment by cloudxxx8** [comment]
hold until 2.0.0 release
this is targeting to 2.0.1

**Review by cloudxxx8 — APPROVED** [review]
LGTM

**Pull request merged**